### PR TITLE
SCIM 2.0 Support for IS 5.3.0 architecture

### PR DESCRIPTION
**Pull request by Vindulamj**

### CODE REVIEW

**Comment by cdwijayarathna** [comment]
Hi,

Can someone give me some idea about this PR? I'm bit confused.
What is 1.0.x branch and how does it defer from master branch?
Why does 1.0.x branch has version 2.0.0-SNAPSHOT? Shoudn't it be 1.0.something?
I see already there is packages with name scim2, then what does this PR do?

**Comment by Vindulamj** [comment]
Hi @cdwijayarathna 

This PR is to add SCIM 2.0 implementation based on C4 user core API. Master branch contains the SCIM 2.0 implementation based on C5 user core API.

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/SCIMGroupHandler.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+import java.util.Set;
+import java.util.UUID;
+
+
+/**
+ * This is for managing SCIM specific attributes related to Group resource in Identity_SCIM_GROUP
+ * Table. This should be managed per tenant.
+ * But need to use the same approach as for User, by going through AttributMapper to do it in a generic way.
+ */
+public class SCIMGroupHandler {
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/SCIMGroupHandler.java]: Why this is inside DAO package?

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/SQLQueries.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.scim.v2.common.DAO;
+
+
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/SQLQueries.java]: Unnecessary new line.

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/SQLQueries.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+public class SQLQueries {
+
+    /*SQL Queries for SCIM_IDENTITY_TABLE which persists SCIM_GROUP info*/
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/SQLQueries.java]: Shouldn't this be the class level comment?

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/impl/IdentitySCIMManager.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            }
+            //get user store manager
+        } catch (UserStoreException e) {
+            String error = "Error obtaining user realm for the user: " + userName;
+            throw new CharonException(error, e);
+        }
+        return scimUserManager;
+    }
+
+
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/impl/IdentitySCIMManager.java]: unnecessary new line

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/impl/SCIMUserManager.java [hunk 5]
```diff
@@ -0,0 +1,1290 @@
...
+import org.wso2.charon3.core.objects.User;
+import org.wso2.charon3.core.schema.SCIMResourceSchemaManager;
+import org.wso2.charon3.core.schema.SCIMResourceTypeSchema;
+import org.wso2.charon3.core.utils.AttributeUtil;
+import org.wso2.charon3.core.utils.ResourceManagerUtil;
+import org.wso2.charon3.core.utils.codeutils.ExpressionNode;
+import org.wso2.charon3.core.utils.codeutils.Node;
+import org.wso2.charon3.core.utils.codeutils.SearchRequest;
+
+import java.util.*;
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/impl/SCIMUserManager.java]: remove * import

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            throws IdentitySCIMException {
+        Connection connection = IdentityDatabaseUtil.getDBConnection();
+        PreparedStatement prepStmt = null;
+
+        if (!isExistingGroup(SCIMCommonUtils.getGroupNameWithDomain(roleName), tenantId)) {
+            try {
+                prepStmt = connection.prepareStatement(SQLQueries.ADD_ATTRIBUTES_SQL);
+                prepStmt.setInt(1, tenantId);
+                prepStmt.setString(2, roleName);
+                for (Map.Entry<String, String> entry : attributes.entrySet()) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java]: Wouldn't it be better if we use SQL bulk insert (batch insert) here?

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        PreparedStatement prepStmt = null;
+
+        if (isExistingGroup(SCIMCommonUtils.getGroupNameWithDomain(roleName), tenantId)) {
+            try {
+                prepStmt = connection.prepareStatement(SQLQueries.UPDATE_ATTRIBUTES_SQL);
+
+                prepStmt.setInt(2, tenantId);
+                prepStmt.setString(3, roleName);
+
+                for (Map.Entry<String, String> entry : attributes.entrySet()) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java]: Can we use SQL batch update?

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/utils/SCIMConfigProcessor.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+import org.apache.axiom.om.OMElement;
+import org.apache.axiom.om.impl.builder.StAXOMBuilder;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.charon3.core.exceptions.CharonException;
+
+import javax.xml.namespace.QName;
+import javax.xml.stream.XMLStreamException;
+import java.io.*;
```
Comment (omindu) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/utils/SCIMConfigProcessor.java]: Don't use * imports.

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/impl/IdentitySCIMManager.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.claim.ClaimManager;
+import org.wso2.carbon.user.core.service.RealmService;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+import org.wso2.charon3.core.config.CharonConfiguration;
+import org.wso2.charon3.core.encoder.JSONEncoder;
+import org.wso2.charon3.core.exceptions.CharonException;
+import org.wso2.charon3.core.extensions.*;
```
Comment (omindu) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/impl/IdentitySCIMManager.java]: Remove * imports.

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/internal/SCIMCommonComponent.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+ * policy="dynamic" bind="setIdentityCoreInitializedEventService" unbind="unsetIdentityCoreInitializedEventService"
+ */
+public class SCIMCommonComponent {
+    private static Log logger = LogFactory.getLog(SCIMCommonComponent.class);
+
+    ExecutorService executorService = Executors.newFixedThreadPool(1);
+
+    protected void activate(ComponentContext ctx) {
+        try {
+            String filePath = CarbonUtils.getCarbonConfigDirPath() + File.separator + "identity" + File.separator +
```
Comment (omindu) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/internal/SCIMCommonComponent.java]: Can use IdentityUtil.getIdentityConfigDirPath()

**Comment by cdwijayarathna** [comment]
In master branch, package names has "scim2", any specific reason to havedifferent name "scim.v2" here? Isn't it good to have the same package structure?

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+        }
+        return isExistingGroup;
+    }
+
+    private boolean isExistingAttribute(String attributeName, String groupName, int tenantId)
+            throws IdentitySCIMException {
+        Connection connection = IdentityDatabaseUtil.getDBConnection();
+        PreparedStatement prepStmt = null;
+        ResultSet rSet = null;
+
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java]: Unnecessary new line

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+                for (Map.Entry<String, String> entry : attributes.entrySet()) {
+                    if (!isExistingAttribute(entry.getKey(),
+                            SCIMCommonUtils.getGroupNameWithDomain(roleName), tenantId)) {
+                        prepStmt.setString(3, entry.getKey());
+                        prepStmt.setString(4, entry.getValue());
+
+                        prepStmt.addBatch();
+                        count++;
+
+                        if (count == attributes.size()) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java]: I think we can get rid of this check and do the executing of the batch outside the for-loop. Similar implementation can be found [here](https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/TokenMgtDAO.java#L1072).

File: components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+                            prepStmt.executeBatch();
+                            connection.commit();
+                        }
+                    } else {
+                        throw new IdentitySCIMException("Error when adding SCIM Attribute: "
+                                + entry.getKey()
+                                + " An attribute with the same name already exists.");
+                    }
+                }
+                connection.commit();
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.scim.v2.common/src/main/java/org/wso2/carbon/identity/scim.v2/common/DAO/GroupDAO.java]: Do we need a connection commit here too?

**Pull request merged**